Objects/typeobject.c

```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
     return 0;
 }
 
+static void
+update_cache(struct type_cache_entry *entry, PyObject *name, unsigned int version_tag, PyObject *value)
+{
+    entry->version = version_tag;
+    entry->value = value;  /* borrowed */
+    assert(_PyASCIIObject_CAST(name)->hash != -1);
+    OBJECT_STAT_INC_COND(type_cache_collisions, entry->name != Py_None && entry->name != name);
+    Py_SETREF(entry->name, Py_NewRef(name));
```

Comment:
The implicit `Py_DECREF` under a lock (the seqlock) worries me, but I guess it's fine here because `name` must be a `PyUnicodeObject` (and not a subclass).